### PR TITLE
Add Type Script declaration file

### PR DESCRIPTION
**Pull request by strausr**
Feature/add ts decleration file

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,1098 @@
...
+
+    export function picture(public_id: string, options?: ImageOptions | UtilsOptions): string;
+
+    export function source(public_id: string, options?: ImageOptions | VideoOptions | UtilsOptions): string;
+
+    export function url(public_id: string, options?: ImageOptions | VideoOptions | UtilsOptions): string;
+
+    export function video(public_id: string, options?: VideoOptions | UtilsOptions): string;
+
+
```
Comment (const-cloudinary) [inline types/index.d.ts]: Pretty title

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,1098 @@
...
+        function upload_preset(name?: string, options?: ApiOptions): Promise<any>;
+
+        function upload_presets(callback?: ErrorCallBack, options?: ApiOptions): Promise<any>;
+
+        function upload_presets(options?: ApiOptions): Promise<any>;
+
+        function usage(callback: ErrorCallBack, options?: ApiOptions): Promise<any>;
+    }
+
+    export namespace uploader {
```
Comment (const-cloudinary) [inline types/index.d.ts]: Pretty title

File: types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,1098 @@
...
+        | "dailymotion";
+    type TargetFormat = string | "zip" | "tgz";
+    type ErrorCallBack = (error: any, result: any) => any;
+
+    // helper functions
+    export function config(new_config: ConfigOptions, new_value?: any): void;
+
+    export function image(source: string, options?: ImageOptions | UtilsOptions): string;
+
+    export function picture(public_id: string, options?: ImageOptions | UtilsOptions): string;
```
Comment (const-cloudinary) [inline types/index.d.ts]: Please add test

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,1098 @@
...
+        function upload_stream(options?: UploadApiOptions): Promise<any>;
+
+        function upload_tag_params(callback?: ErrorCallBack, options?: UploadApiOptions): Promise<any>;
+
+        function upload_tag_params(options?: UploadApiOptions): Promise<any>;
+
+        function upload_url(options?: UploadApiOptions): Promise<any>;
+    }
+
+    export namespace utils {
```
Comment (const-cloudinary) [inline types/index.d.ts]: move closer to helpers

File: types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,1098 @@
...
+
+            function create_archive(options?: UploadApiOptions, target_format?: TargetFormat, callback?: ErrorCallBack,): Promise<any>;
+
+            function create_zip(options?: UploadApiOptions, callback?: ErrorCallBack): Promise<any>;
+
+            function destroy(public_id: string, options?: { resource_types?: ResourceType, type?: DeliveryType, invalidate?: boolean }, callback?: ErrorCallBack,): Promise<any>;
+
+            function destroy(public_id: string, callback?: ErrorCallBack,): Promise<any>;
+
+            function direct_upload(callback_url: string, options?: UploadApiOptions): Promise<any>;
```
Comment (const-cloudinary) [inline types/index.d.ts]: add test to fix other unused

File: types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,1098 @@
...
+
+            function video_url(public_id?: string, options?: VideoOptions): string;
+
+            function webhook_signature(data?: string, timestamp?: number, options?: any): string;
+
+            function zip_download_url(tag?: string, options?: ApiOptions): string;
+        }
+    }
+
+    export interface CommonTransformationOptions {
```
Comment (const-cloudinary) [inline types/index.d.ts]: add future key

File: types/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,1098 @@
...
+        width?: string | number; // Number of pixels, width % or "auto" with rounding step
```
Comment (const-cloudinary) [inline types/index.d.ts]: Please remove comments and optionally add them to the actual code

File: types/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,1098 @@
...
+        delay?: number | string;
+        quality?: string | number; // percent or percent[:chroma_subsampling] or auto[:quality_level]
+        dpr?: number | string; // Deliver the image in the specified device pixel ratio. The parameter accepts any positive float value
+        if?: string; // Apply a transformation only if a specified condition is met (see the conditional transformations documentation).
+        variables?: Array<[string, any]>;
+        transformation?: string | Array<ImageOptions> | Array<VideoOptions>; // Apply a pre-defined named transformation of the given name. When using Cloudinary's client integration libraries, the 'transformation' parameter accepts an array of transformation parameters to be chained together.
+        effect?: string | Array<string | number>; // name and value, e.g. hue:40
+        radius?: number | string; // pixels or max
+        custom_function?: string | { function_type: FunctionType, source: string }
+        sign_url?: boolean;
```
Comment (const-cloudinary) [inline types/index.d.ts]: move to relevant options

File: types/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,1098 @@
...
+    }
+
+    export interface UtilsOptions extends CommonTransformationOptions {
```
Comment (const-cloudinary) [inline types/index.d.ts]: Separate to url and utils(helper) options.

UrlOptions extends ConfigOptions

File: types/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,1098 @@
...
+        responsive_placeholder?: boolean;
+        source_types?: string | string[];
+        source_transformation?: string;
+        poster?: string | object;
+        fallback_content?: string;
+        unsafe_update?: string;
+        allowed_for_strict?: boolean;
+    }
+
+    export interface ImageOptions extends CommonTransformationOptions {
```
Comment (const-cloudinary) [inline types/index.d.ts]: Rename to ImageTransformationOptions

File: types/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,1098 @@
...
+    export interface CommonTransformationOptions {
+        width?: string | number; // Number of pixels, width % or "auto" with rounding step
+        height?: number | string; // Number of pixels or height %
+        crop?: CropMode;
+        aspect_ratio?: string | number | string; // ratio or percent, e.g. 1.5 or 16:9
+        gravity?: Gravity; // The last any covers auto:50 which is cropping algorithm aggresiveness and future proofing
+        background?: string; // color, e.g. "blue" or "rgb:9090ff"
+        overlay?: string | Object; // Identifier, e.g. "text:Arial_50:Smile!", or public id of a different
+        angle?: Angle; // degrees or mode
+        delay?: number | string;
```
Comment (const-cloudinary) [inline types/index.d.ts]: move to video

File: types/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,1098 @@
...
+        source_transformation?: string;
+        poster?: string | object;
+        fallback_content?: string;
+        unsafe_update?: string;
+        allowed_for_strict?: boolean;
+    }
+
+    export interface ImageOptions extends CommonTransformationOptions {
+        border?: string; // style, e.g. "6px_solid_rgb:00390b60"
+        color?: string; // e.g. "red" or "rgb:20a020"
```
Comment (const-cloudinary) [inline types/index.d.ts]: move to common

File: types/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,1098 @@
...
+        color_space?: ColorSpace;
+        default_image?: string; // public id of an uploaded image
+        density?: number | string; // Control the density to use while converting a PDF document to images. (range: 50-300, default: 150)
+        fetch_format?: ImageFileExtension;
+        format?: ImageFileExtension;
+        flags?: ImageFlags | string; // Set one or more flags that alter the default transformation behavior. Separate multiple flags with a dot (`.`).
+        else?: string;
```
Comment (const-cloudinary) [inline types/index.d.ts]: move to common

File: types/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,1098 @@
...
+        end_if?: string;
+        opacity?: number | string; // percent, 0-100
+        page?: number | string; // Given a multi-page file (PDF, animated GIF, TIFF), generate an image of a single page using the given index.
+        raw_transformation?: any;
```
Comment (const-cloudinary) [inline types/index.d.ts]: common

File: types/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,1098 @@
...
+        size?: string;
```
Comment (const-cloudinary) [inline types/index.d.ts]: common

File: types/index.d.ts [hunk 16]
```diff
@@ -0,0 +1,1098 @@
...
+        bit_rate?: number | string; // Advanced control of video bitrate in bits per second. By default the video uses a variable bitrate (VBR), with this value indicating the maximum bitrate. If constant is specified, the video plays with a constant bitrate (CBR).
+        // Supported codecs: h264, h265(MPEG - 4); vp8, vp9(WebM).
+        duration?: number | string; // Float or string
+        end_offset?: number | string; // Float or string
+        fallbackContent?: string;
+        flags?: VideoFlags;
+        fps?: string | Array<string | number>;
+        keyframe_interval?: number;
+        offset?: string, // [float, float] or [string, string] or a range. Shortcut to set video cutting using a combination of start_offset and end_offset values
+        poster?: string | Object,
```
Comment (const-cloudinary) [inline types/index.d.ts]: move to utils(helpers)

File: types/index.d.ts [hunk 17]
```diff
@@ -0,0 +1,1098 @@
...
+        source_transformation?: string;
+        start_offset?: number | string; // Float or string
+        streaming_profile?: StreamingProfiles
+        video_codec?: string | Object; // Select the video codec and control the video content of the profile used. Can be provided in the form <codec>[:<profile>:[<level>]] to specify specific values to apply for video codec, profile and level, e.g. "h264:baseline:3.1". Also accepts a hash of values such as { codec: 'h264', profile: 'basic', level: '3.1' }
+        video_sampling?: number | string; // Integer - The total number of frames to sample from the original video. The frames are spread out over the length of the video, e.g. 20 takes one frame every 5% -- OR -- String - The number of seconds between each frame to sample from the original video. e.g. 2.3s takes one frame every 2.3 seconds.
+        format?: VideoFileExtension;
+        fetch_format?: VideoFileExtension;
+        preload?: string;
+        fallback_content?: string;
+        controls?: boolean;
```
Comment (const-cloudinary) [inline types/index.d.ts]: tag(utils)

File: types/index.d.ts [hunk 18]
```diff
@@ -0,0 +1,1098 @@
...
+        source_types?: string;
+        source_transformation?: string;
+        start_offset?: number | string; // Float or string
+        streaming_profile?: StreamingProfiles
+        video_codec?: string | Object; // Select the video codec and control the video content of the profile used. Can be provided in the form <codec>[:<profile>:[<level>]] to specify specific values to apply for video codec, profile and level, e.g. "h264:baseline:3.1". Also accepts a hash of values such as { codec: 'h264', profile: 'basic', level: '3.1' }
+        video_sampling?: number | string; // Integer - The total number of frames to sample from the original video. The frames are spread out over the length of the video, e.g. 20 takes one frame every 5% -- OR -- String - The number of seconds between each frame to sample from the original video. e.g. 2.3s takes one frame every 2.3 seconds.
+        format?: VideoFileExtension;
+        fetch_format?: VideoFileExtension;
+        preload?: string;
+        fallback_content?: string;
```
Comment (const-cloudinary) [inline types/index.d.ts]: tag

File: types/index.d.ts [hunk 19]
```diff
@@ -0,0 +1,1098 @@
...
+        effect?: string | Array<string | number> | VideoEffect; // name and value, e.g. hue:40
+    }
+
+    interface ConfigOptions {
+        cloud_name?: string;
+        api_key?: string;
+        api_secret?: string;
+        secure?: boolean;
+        upload_preset?: string;
+        upload_prefex?: string;
```
Comment (const-cloudinary) [inline types/index.d.ts]: no such word

File: types/index.d.ts [hunk 20]
```diff
@@ -0,0 +1,1098 @@
...
+        controls?: boolean;
+        effect?: string | Array<string | number> | VideoEffect; // name and value, e.g. hue:40
+    }
+
+    interface ConfigOptions {
+        cloud_name?: string;
+        api_key?: string;
+        api_secret?: string;
+        secure?: boolean;
+        upload_preset?: string;
```
Comment (const-cloudinary) [inline types/index.d.ts]: move to apiOptions

File: types/index.d.ts [hunk 21]
```diff
@@ -0,0 +1,1098 @@
...
+        upload_prefex?: string;
+        cdn_subdomain?: boolean;
+        private_cdn?: boolean;
+        cname?: string;
+        secure_distribution?: string;
+    }
+
+    export interface ApiOptions  extends ConfigOptions{
```
Comment (const-cloudinary) [inline types/index.d.ts]: Rename to CommonApiOptions

File: types/index.d.ts [hunk 22]
```diff
@@ -0,0 +1,1098 @@
...
+        bytes_step?: number;
+        max_images?: number;
+        to_type?: string;
+        command?: string;
+        prefixes?: string;
+        transformations?: string;
+        mode?: string;
+        target_format?: string;
+        target_public_id?: string;
+        flatten_folders?: boolean;
```
Comment (const-cloudinary) [inline types/index.d.ts]: move to ArchiveApiOptions

**Comment by strausr** [comment]
@FelixZilber Thanks! We'll take this into consideration in the next phase

File: types/cloudinary-tests.ts [hunk 23]
```diff
@@ -0,0 +1,1152 @@
...
+    });
+
+// $ExpectType Promise<any>
+cloudinary.v2.api.delete_upload_mapping('wiki',
+    function (error, result) {
+        console.log(result);
+    });
+
+// $ExpectError
+cloudinary.api.delete_upload_preset(123,
```
[the remark this answers is not in the document]
Comment (strausr) [reply types/cloudinary-tests.ts]: Can't use  // @ts-ignore with dtslint

File: README.md [hunk 24]
```diff
@@ -202,6 +202,11 @@ cloudinary.image("sample", {format: "png", width: 100, height: 100, crop: "fill"
 // <img src='http://res.cloudinary.com/demo/image/upload/c_fill,h_100,w_100/sample.png' height='100' width='100'/>
 ```
 
+### Typescript
+
+We added support for TS, The SDK can now be consumed in a TS environment.
+You can view the decleration file [here](https://github.com/cloudinary/cloudinary_npm/types/index.d.ts).  
+
```
Comment (asisayag2) [inline README.md]: Would rephrase to:
"🎉New 🎉TypeScript support was just added. Check out the [declaration file] (URL) "
